### DISCUSSION TITLE
What may be causing custom server config to NOT be utilized? (Omnisharp on Linux)

### DISCUSSION THREAD

**Discussion by Blake-LeBlanc** [mason-org/mason.nvim#232] [Q&A]
(This question may relate more to `mason-lspconfig.nvim`, the issue form on that repo directed me here)

I've noticed that regardless of what I include in the custom `omnisharp.lua` file, the configuration outlined in `:LspInfo` seems to indicate the default `cmd` that is cobbled together in `nvim-lspconfig`. 

I suspect there's something I'm not doing right in setting up `mason.nvim` and `mason-lspconfig.nvim` since the `cmd` output updates when I use `williamboman/nvim-lsp-installer`.

That said, I haven't been able to get omnisharp to work in either case. This is on Linux (Ubuntu 20.04.4), trying to get it up and running for Unity C# Mono scripts. 

Any ideas where I might be going wrong? 

As you can see in the code below in the `configs.lua` file, I've tried making using the omnisharp-roslyn that comes with `:MasonInstall` as well as a direct reference to the mono complete package I installed separately per the instructions at https://www.mono-project.com/download/stable/ . 

I wouldn't be surprised if I need to somehow integrate Mason's `omnisharp-mono` server rather than `omnisharp`, but whenever I try that route, it rightly complains that `nvim-lspconfig` does not have a corresponding "match" to refer to.

Thanks in advance for any help!

#### /home/linux/dotfiles/neovim/.config/nvim/lua/user/lsp/configs.lua
```
local status_ok, lsp_installer = pcall(require, "mason")
-- local status_ok, mason = pcall(require, "mason")
if not status_ok then
	return
end

local lspconfig = require("lspconfig")
local mason_lspconfig = require("mason-lspconfig")

-- https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md
-- https://github.com/williamboman/mason.nvim/blob/main/lua/mason-registry/index.lua

local servers = {
  "bashls",
  "omnisharp",
    -- FIXME: How to set this up for the mono version in mason?
    -- "omnisharp-mono",
  "sumneko_lua",
  "tsserver",

}

lsp_installer.setup({
  ensure_installed = servers,

})

for _, server in pairs(servers) do
  local opts = {
    on_attach = require("user.lsp.handlers").on_attach,
    capabilities = require("user.lsp.handlers").capabilities,
  }
  
  local has_custom_opts, server_custom_opts = pcall(require, "user.lsp.settings." .. server)
  if has_custom_opts then
    opts = vim.tbl_deep_extend("force", opts, server_custom_opts)
  end

  -- NOTE: This is what was used with `nvim-lsp-installer`
  -- lspconfig[server].setup(opts)

  -- NOTE: This is what I'm attempting to use for `mason` and `mason-lspconfig`
  mason_lspconfig.setup({
    lspconfig[server].setup(opts)
  })

end
```

#### plugins
```
use('neovim/nvim-lspconfig')
use('williamboman/mason.nvim')
use('williamboman/mason-lspconfig.nvim')
```

#### Treemap of directory structure, just to help give some context
And just ignore those extra `omnisharp_MASON.lua`, etc files, those are just where I'm backing up my various trial-and-error attempts

![Screenshot_2022-08-02_16-31-01](https://user-images.githubusercontent.com/19244245/182483288-8a659602-3783-4194-9e6a-18dea4aa35ec.png)

#### lsp/settings/omnisharp.lua
```
return {
  config = {
    -- cmd = { "mono", "/home/linux/.local/share/nvim/mason/packages/omnisharp/OmniSharp.Roslyn.dll" },
    cmd = { "mono", "/bin/mono" },
    use_modern_net = false,

    --- From defaults
    enable_editorconfig_support = true,
    enable_roslyn_analyzers = false,
    organize_imports_on_format = false,
    enable_import_completion = false,
    sdk_include_prereleases = true,
    analyze_open_documents_only = false,
    filetypes = { 'cs', 'vb' },

  },
}

```
#### :LspInfo
```
 1 client(s) attached to this buffer: 
 
 Client: omnisharp (id: 5, pid: 12456, bufnr: [119])
 	filetypes:       cs, vb
 	autostart:       true
 	root directory:  /home/linux/<redacted>
 	cmd:             omnisharp -z -s /home/linux/<redacted> --hostPID 1627 DotNet:enablePackageRestore=false --encoding utf-8 --languageserver FormattingOptions:EnableEditorConfigSupport=true Sdk:IncludePrereleases=true
```

That `cmd` output seems to come from how the `cmd` is built in the following...
 
#### /nvim-lspconfig/lua/lspconfig/server_configurations/omnisharp.lua
  ```
  ...

  on_new_config = function(new_config, new_root_dir)
      table.insert(new_config.cmd, '-z') -- https://github.com/OmniSharp/omnisharp-vscode/pull/4300
      vim.list_extend(new_config.cmd, { '-s', new_root_dir })
      vim.list_extend(new_config.cmd, { '--hostPID', tostring(vim.fn.getpid()) })
      table.insert(new_config.cmd, 'DotNet:enablePackageRestore=false')
      vim.list_extend(new_config.cmd, { '--encoding', 'utf-8' })
      table.insert(new_config.cmd, '--languageserver')

      if new_config.enable_editorconfig_support then
        table.insert(new_config.cmd, 'FormattingOptions:EnableEditorConfigSupport=true')
      end

      if new_config.organize_imports_on_format then
        table.insert(new_config.cmd, 'FormattingOptions:OrganizeImports=true')
      end

      if new_config.enable_ms_build_load_projects_on_demand then
        table.insert(new_config.cmd, 'MsBuild:LoadProjectsOnDemand=true')
      end

      if new_config.enable_roslyn_analyzers then
        table.insert(new_config.cmd, 'RoslynExtensionsOptions:EnableAnalyzersSupport=true')
      end

      if new_config.enable_import_completion then
        table.insert(new_config.cmd, 'RoslynExtensionsOptions:EnableImportCompletion=true')
      end

      if new_config.sdk_include_prereleases then
        table.insert(new_config.cmd, 'Sdk:IncludePrereleases=true')
      end

      if new_config.analyze_open_documents_only then
        table.insert(new_config.cmd, 'RoslynExtensionsOptions:AnalyzeOpenDocumentsOnly=true')
      end
    end,
  ...
```

**Comment by williamboman** [comment]
Hello! So the duality of Omnisharp's different runtimes (Mono vs .NET) is a bit messy. By default, the .NET variant will be used if you request `omnisharp` to be installed.

If you want to use the Mono variant, I'd recommend doing the following:

- Separate `omnisharp` from your other lspconfig servers (i.e., remove from the `ensure_installed` table, and set it up separately outside of the `for` loop)
- Install the `omnisharp-mono` package (manually or automated via [mason-tool-installer.nvim](https://github.com/WhoIsSethDaniel/mason-tool-installer.nvim))
- When setting up `omnisharp` via `lspconfig`, set the `use_modern_net = false` flag:

```lua
local servers = {
  "bashls",
  "sumneko_lua",
  "tsserver",

}

mason_lspconfig.setup({
  ensure_installed = servers,
})

for _, server in pairs(servers) do
  local opts = {
    on_attach = require("user.lsp.handlers").on_attach,
    capabilities = require("user.lsp.handlers").capabilities,
  }
  
  local has_custom_opts, server_custom_opts = pcall(require, "user.lsp.settings." .. server)
  if has_custom_opts then
    opts = vim.tbl_deep_extend("force", opts, server_custom_opts)
  end

  lspconfig[server].setup(opts)
end

lspconfig.omnisharp.setup {
  use_modern_net = false
}
```

I also noticed you're calling:

```lua
  mason_lspconfig.setup({
    lspconfig[server].setup(opts)
  })
```

It's a bit of a weird expression that I'm not gonna describe in detail - but, remove it :D

**Reply by Blake-LeBlanc** [reply]
> but, remove it :D

Done, gone :)

---

> set the use_modern_net = false flag

Cool, got it!

---

Mason installs the server just fine, but it still seems to be ignoring any custom `cmd` value.

Which, maybe I'm misunderstanding something (entirely possible!), but in addition to setting the `use_modern_net` and `use_mono` flags, isn't being able to set a custom `cmd` value also a requirement to getting a Linux+Mono config up and running?

It seems like, with `mason` and `mason-lspconfig`, the `cmd` value is still being ignored in favor of the `on_new_config` value.

Let me try to illustrate it more directly...

After separating omnisharp's `.setup` per your recommendation...

With the previous nvim-lsp-installer, if I set `cmd` to something like "TESTING", it *WILL* respect and attempt to use this value

#### configs.lua
```
local status_ok, lsp_installer = pcall(require, "nvim-lsp-installer")
if not status_ok then
	return
end

local lspconfig = require("lspconfig")

local servers = {
  "bashls",
  "sumneko_lua",

}

lsp_installer.setup({
	ensure_installed = servers,

})

for _, server in pairs(servers) do
	local opts = {
		on_attach = require("user.lsp.handlers").on_attach,
		capabilities = require("user.lsp.handlers").capabilities,
	}

	local has_custom_opts, server_custom_opts = pcall(require, "user.lsp.settings." .. server)

	if has_custom_opts then
		opts = vim.tbl_deep_extend("force", opts, server_custom_opts)
	end

	lspconfig[server].setup(opts)

end

-- omnisharp-mono directory
local omnisharp_bin = "mono /home/linux/.local/share/nvim/lsp_servers/omnisharp/omnisharp-mono/OmniSharp.exe"

... 

local pid = vim.fn.getpid()
local util = require('lspconfig/util')

lspconfig.omnisharp.setup({
  on_attach = require("user.lsp.handlers").on_attach,
  capabilities = require("user.lsp.handlers").capabilities,

  use_mono = true,
  use_modern_net = false,

  -- cmd = { omnisharp_bin, "--languageserver" , "--hostPID", tostring(pid) },
  cmd = { "TESTING" },

  enable_editorconfig_support = true,
  enable_ms_build_load_projects_on_demand = false,
  enable_roslyn_analyzers = true,
  organize_imports_on_format = false,
  enable_import_completion = false,
  sdk_include_prereleases = true,
  analyze_open_documents_only = false,
  root_dir = util.root_pattern('*.sln'),
  filetypes = { 'cs', 'vb' },

})
```

As expected, when entering into a `*.cs` file it will complain about the `cmd` value:

```
Spawning language server with cmd: `TESTING` failed. The language server is either not installed, missing from PATH, or not executable.
```

---

Now, whenever I swap things out for `mason` and `mason-lspconfig`, this does not happen. The custom `cmd` value is not used. Neither when it's "TESTING" or as the more practical value...

#### configs.lua
```
local status_ok, lsp_installer = pcall(require, "mason")
if not status_ok then
	return
end

local lspconfig = require("lspconfig")
require("mason-lspconfig").setup()


local servers = {
  "bashls",
  "sumneko_lua",

}

lsp_installer.setup({
	ensure_installed = servers,

})

for _, server in pairs(servers) do
	local opts = {
		on_attach = require("user.lsp.handlers").on_attach,
		capabilities = require("user.lsp.handlers").capabilities,
	}

	local has_custom_opts, server_custom_opts = pcall(require, "user.lsp.settings." .. server)

	if has_custom_opts then
		opts = vim.tbl_deep_extend("force", opts, server_custom_opts)
	end

	lspconfig[server].setup(opts)

end

-- omnisharp-mono directory
-- local omnisharp_bin = "mono /home/linux/.local/share/nvim/lsp_servers/omnisharp/omnisharp-mono/OmniSharp.exe"
local omnisharp_bin = "mono /home/linux/.local/share/nvim/mason/packages/omnisharp-mono/omnisharp/OmniSharp.exe"
  -- https://github.com/williamboman/nvim-lsp-installer/issues/479#issuecomment-1128840405

... 

local pid = vim.fn.getpid()
local util = require('lspconfig/util')

lspconfig.omnisharp.setup({
  on_attach = require("user.lsp.handlers").on_attach,
  capabilities = require("user.lsp.handlers").capabilities,

  use_mono = true,
  use_modern_net = false,

  -- cmd = { omnisharp_bin, "--languageserver" , "--hostPID", tostring(pid) },
  cmd = { "TESTING" },

  enable_editorconfig_support = true,
  enable_ms_build_load_projects_on_demand = false,
  enable_roslyn_analyzers = true,
  organize_imports_on_format = false,
  enable_import_completion = false,
  sdk_include_prereleases = true,
  analyze_open_documents_only = false,
  root_dir = util.root_pattern('*.sln'),
  filetypes = { 'cs', 'vb' },

})
```

#### :LspInfo
```
Config: omnisharp
 	filetypes:         cs, vb
 	root directory:    /home/linux/<redacted>
 	cmd:               omnisharp-mono -z -s /home/linux/<redacted> --hostPID 45924 DotNet:enablePackageRestore=false --encoding utf-8 --languageserver FormattingOptions:EnableEditorConfigSupport=true RoslynExtensionsOptions:EnableAnalyzersSupport=true Sdk:IncludePrereleases=true
 	cmd is executable: true
 	autostart:         true
 	custom handlers:   
```

Thanks again for your help @williamboman !

**Reply by williamboman** [reply]
First, thanks for the detailed report! It helps me tremendously, I'll use yours as an example of how to do it perfectly haha.

I just realized that the `cmd` is always overwritten (if a custom one exists in `mason-lspconfig`). It's definitely not by design (user options should always take precedence), but more a result of some limitations. I'll look into that, it will likely involve minor changes in lspconfig itself. This has actually been the case for a long time, including in `nvim-lsp-installer`. The reason Omnisharp behaves differently between the two plugins is that `mason-lspconfig` provides [a custom cmd](https://github.com/williamboman/mason-lspconfig.nvim/blob/main/lua/mason-lspconfig/server_configurations/omnisharp/init.lua#L3).

> Which, maybe I'm misunderstanding something (entirely possible!), but in addition to setting the use_modern_net and use_mono flags, isn't being able to set a custom cmd value also a requirement to getting a Linux+Mono config up and running?

I can't really see any reasons for changing the `cmd` for Omnisharp. The docs in lspconfig suggest that it is necessary, but `mason` takes care of that for you. Looking at the `:LspInfo` output at the end, it seems fine to me. Is the server not working with that setup?

**Reply by Blake-LeBlanc** [reply]
> First, thanks for the detailed report! It helps me tremendously

Glad to hear it, you're welcome!

> I can't really see any reasons for changing the cmd for Omnisharp. The docs in lspconfig suggest that it is necessary, but mason takes care of that for you. 

Oh very cool! That's good to know, awesome change!

> Looking at the :LspInfo output at the end, it seems fine to me. Is the server not working with that setup?

No, it doesn't seem to ever attach. The full-blown `:LspInfo` output confirms `0 client(s) to this buffer:` (full output at the bottom of this post)

Before I dive into that though, let me take a quick moment to clarify something on my end.

Looking back over your earlier post, I saw you had...
```
mason_lspconfig.setup({
  ensure_installed = servers,
})
```

Whenever I have it set up that way, none of the installed servers are recognized. When I open a related file, neovim immediately messages;
```
Spawning language server with cmd: `<server>` failed. The language server is either not installed, missing from PATH, or not executable.
```

However, when I use the NON-masonized version: 
```
lspconfig.setup({
  ensure_installed = servers,
})
```
The named servers are discovered and work okay (lua, bash, typescript, etc).

But this leaves me with an "empty" `mason_lspconfig.setup()` call just sitting there in the config doing nothing.

I only kick-off with that just in case maybe there's something there on my end that's preventing the omnisharp/omnisharp-mono servers from working their magic.

---

Okay, so to dive back into omnisharp... 

Immediately upon loading a `*.cs` file, the following message displays across the bottom

```
Client 1 quit with exit code 2 and signal 0
```

So I pull up the logs and see...

#### :LspLog
```
[START][2022-08-10 22:01:26] LSP logging initiated
[ERROR][2022-08-10 22:01:26] .../vim/lsp/rpc.lua:439	"rpc"	"omnisharp-mono"	"stderr"	"Cannot open assembly '/home/linux/.local/share/nvim/mason/bin/omnisharp/OmniSharp.exe': Not a directory.\n"
```

Which actually seems consintent with mason's current (new?) directory structure. Because while it's looking for an omnisharp directory within `/bin`, it now looks to be laid out differently 

#### $ ls
```
(jobs:0) linux:/home/linux/.local/share/nvim/mason/bin 
$ ls
total 12
drwxr-xr-x 2 linux linux 4096 Aug 10 21:54 .
drwxr-xr-x 5 linux linux 4096 Jul 28 16:36 ..
lrwxrwxrwx 1 linux linux   71 Aug 10 21:53 bash-language-server -> ../packages/bash-language-server/node_modules/.bin/bash-language-server
lrwxrwxrwx 1 linux linux   51 Aug 10 21:53 lua-language-server -> ../packages/lua-language-server/lua-language-server
lrwxrwxrwx 1 linux linux   31 Aug 10 21:54 omnisharp -> ../packages/omnisharp/omnisharp
lrwxrwxrwx 1 linux linux   30 Aug 10 21:54 omnisharp-mono -> ../packages/omnisharp-mono/run
```

And as we'd expect, `:LspInfo` shows it unattached
#### :LspInfo
```
 Language client log: /home/linux/.local/state/nvim/lsp.log
 Detected filetype:   cs
 
 0 client(s) attached to this buffer: 
 
 2 active client(s) not attached to this buffer: 
 
 Client: sumneko_lua (id: 2, pid: 68593, bufnr: [17, 16, 8])
 	filetypes:       lua
 	autostart:       true
 	root directory:  /home/linux/dotfiles
 	cmd:             lua-language-server
 
 Client: null-ls (id: 3, pid: 5001, bufnr: [17, 16, 8])
 	filetypes:       markdown, yaml, html, less, typescriptreact, javascriptreact, scss, javascript, css, typescript, graphql, vue, jsonc, json, handlebars, python, lua
 	autostart:       false
 	root directory:  /home/linux/dotfiles
 	cmd:             nvim
 
 Other clients that match the filetype: cs
 
 Config: omnisharp
 	filetypes:         cs, vb
 	root directory:    /home/linux/<redacted>
 	cmd:               omnisharp-mono -z -s /home/linux/<redacted> --hostPID 68548 DotNet:enablePackageRestore=false --encoding utf-8 --languageserver FormattingOptions:EnableEditorConfigSupport=true RoslynExtensionsOptions:EnableAnalyzersSupport=true Sdk:IncludePrereleases=true -z -s /home/linux/<redacted> --hostPID 68548 DotNet:enablePackageRestore=false --encoding utf-8 --languageserver FormattingOptions:EnableEditorConfigSupport=true RoslynExtensionsOptions:EnableAnalyzersSupport=true Sdk:IncludePrereleases=true
 	cmd is executable: true
 	autostart:         true
 	custom handlers:   
 
 Configured servers list: sumneko_lua, omnisharp, bashls
```

Which, just for comparison, is different from what happens with `nvim-lsp-installer`. There, it will attach, the keymaps defined in `handlers.lua` work, it just won't do anything useful. Everything returns with "No information available".

#### :LspInfo with nvim-lsp-installer
```
 Language client log: /home/linux/.local/state/nvim/lsp.log
 Detected filetype:   cs
 
 1 client(s) attached to this buffer: 
 
 Client: omnisharp (id: 1, pid: 78287, bufnr: [4])
 	filetypes:       cs, vb
 	autostart:       true
 	root directory:  /home/linux/<redacted>
 	cmd:             mono /home/linux/.local/share/nvim/lsp_servers/omnisharp/omnisharp-mono/OmniSharp.exe --languageserver --hostPID 78259
 
 Configured servers list: omnisharp, bashls, sumneko_lua
```
---
---
EDIT: And just to give some additional context, I have BOTH `omnisharp` and `omnisharp-mono` installed through Mason. If I ever try to run with JUST `omnisharp-mono`, I'm quickly met with the following message upon opening  a related file

```
Error detected while processing BufReadPost Autocommands for "*":                                                                                                                        
Error executing lua callback: /usr/local/share/nvim/runtime/filetype.lua:20: Vim(lua):[lspconfig] cmd not defined for "omnisharp". Manually set cmd in the setup {} call according to server_confi
gurations.md, see :help lspconfig-index.
stack traceback:
        [C]: in function 'nvim_buf_set_option'
        /usr/local/share/nvim/runtime/filetype.lua:20: in function </usr/local/share/nvim/runtime/filetype.lua:11>
```
Once I `:MasonInstall omnisharp` and restart, that error goes away and we're back at where we started this post.

Thanks again!

**Reply by williamboman** [reply]
I saw you were using some old variable names like `lsp_installer` for `require("mason")`, so I'm tbh not sure what variable is what in your example 😬. `ensure_installed` should work fine.

I ran into more issues with the `omnisharp-mono`, see:
- https://github.com/williamboman/mason-lspconfig.nvim/pull/32
- https://github.com/williamboman/mason.nvim/pull/290

Could you try pulling the latest plugin versions and instead of setting up `omnsiharp`, set up `omnisharp_mono` instead? This should also now be supported by `ensure_installed`, so you can revert back to your earlier variant where you loop through servers:

```lua
local servers = {
  "bashls",
  "sumneko_lua",
  "tsserver",
  "omnisharp_mono",
}

mason_lspconfig.setup({
  ensure_installed = servers,
})

for _, server in pairs(servers) do
  local opts = {
    on_attach = require("user.lsp.handlers").on_attach,
    capabilities = require("user.lsp.handlers").capabilities,
  }
  
  local has_custom_opts, server_custom_opts = pcall(require, "user.lsp.settings." .. server)
  if has_custom_opts then
    opts = vim.tbl_deep_extend("force", opts, server_custom_opts)
  end

  lspconfig[server].setup(opts)
end
```

**Reply by williamboman** [reply]
See https://github.com/williamboman/mason-lspconfig.nvim/blob/main/lua/mason-lspconfig/server_configurations/omnisharp/README.md for further explanatio!

**Reply by Blake-LeBlanc** [reply]
> Could you try pulling the latest plugin versions and instead of setting up omnsiharp, set up omnisharp_mono instead? This should also now be supported by ensure_installed, so you can revert back to your earlier variant where you loop through servers

Okay, I'll try tomorrow and report back!

**Reply by Blake-LeBlanc** [reply]
@williamboman Good news, thanks to your recent updates, the `omnisharp_mono` LSP is attaching! :)

```
 Language client log: /home/linux/.local/state/nvim/lsp.log
 Detected filetype:   cs
 
 1 client(s) attached to this buffer: 
 
 Client: omnisharp_mono (id: 1, pid: 6582, bufnr: [19])
 	filetypes:       cs, vb
 	autostart:       true
 	root directory:  <redacted>
 	cmd:             omnisharp-mono -z -s <redacted> --hostPID 6566 DotNet:enablePackageRestore=false --encoding utf-8 --languageserver FormattingOptions:EnableEditorConfigSupport=true Sdk:IncludePrereleases=true
 
 Configured servers list: omnisharp_mono, bashls, sumneko_lua
```

`:LspLog` does report issues, but from what I read and after trying to resolve it over the weekend, it seems like that's something on my end not related to Mason.

```
[ERROR][2022-08-15 14:41:35] .../vim/lsp/rpc.lua:439	"rpc"	"omnisharp-mono"	"stderr"	"﻿"
[WARN][2022-08-15 14:41:37] ...lsp/handlers.lua:485	"OmniSharp.CompositionHostBuilder: It looks like you have Mono installed which contains a MSBuild lower than 17.0.0 which is the minimum supported by the configured .NET Core Sdk.\n Try updating Mono to the latest stable or preview version to enable better .NET Core Sdk support. | "
```

Thank you again for all your help William, I appreciate everything you did to help get it this far!

*high fives*
